state/mysql/mysql.go

```diff
@@ -0,0 +1,441 @@
+	// two rows affected
+	return m.returnNDBResults(result, err, 2)
+}
+
+// BulkSet adds/updates multiple entities on store
+// Store Interface
+func (m *MySQL) BulkSet(req []state.SetRequest) error {
+	return m.executeMulti(req, nil)
+}
+
+// Multi handles multiple transactions.
+// TransactionalStore Interface
+func (m *MySQL) Multi(request *state.TransactionalStateRequest) error {
+	var sets []state.SetRequest
+	var deletes []state.DeleteRequest
+
+	for _, req := range request.Operations {
+		switch req.Operation {
+		case state.Upsert:
+			setReq, ok := req.Request.(state.SetRequest)
+
+			if ok {
+				sets = append(sets, setReq)
+			} else {
+				return fmt.Errorf("expecting set request")
+			}
+
+		case state.Delete:
+			delReq, ok := req.Request.(state.DeleteRequest)
+
+			if ok {
+				deletes = append(deletes, delReq)
+			} else {
+				return fmt.Errorf("expecting delete request")
+			}
+
+		default:
+			return fmt.Errorf("unsupported operation: %s", req.Operation)
+		}
+	}
+
+	if len(sets) > 0 || len(deletes) > 0 {
+		return m.executeMulti(sets, deletes)
+	}
+
+	return nil
+}
+
+// BulkGet performs a bulks get operations
+func (m *MySQL) BulkGet(req []state.GetRequest) (bool, []state.BulkGetResponse, error) {
+	// by default, the store doesn't support bulk get
+	// return false so daprd will fallback to call get() method one by one
+	return false, nil, nil
+}
+
+// Close implements io.Closer
+func (m *MySQL) Close() error {
+	if m.db != nil {
+		return m.db.Close()
+	}
+
+	return nil
+}
+
+func (m *MySQL) executeMulti(sets []state.SetRequest, deletes []state.DeleteRequest) error {
+	m.logger.Debug("Executing multiple MySql operations")
+
+	tx, err := m.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	if len(deletes) > 0 {
+		for _, d := range deletes {
+			da := d // Fix for goSec G601: Implicit memory aliasing in for loop.
+			err = m.Delete(&da)
+			if err != nil {
+				tx.Rollback()
+
+				return err
+			}
+		}
+	}
+
+	if len(sets) > 0 {
+		for _, s := range sets {
+			sa := s // Fix for goSec G601: Implicit memory aliasing in for loop.
+			err = m.Set(&sa)
+			if err != nil {
+				tx.Rollback()
+
+				return err
+			}
+		}
+	}
+
+	err = tx.Commit()
+
+	return err
+}
+
+// Verifies that the sql.Result affected no more than n number of rows and no
+// errors exist. If zero rows were affected something is wrong and an error
+// is returned.
+func (m *MySQL) returnNDBResults(result sql.Result, err error, n int64) error {
+	if err != nil {
+		m.logger.Debug(err)
+
+		return err
+	}
+
+	rowsAffected, resultErr := result.RowsAffected()
+
+	if resultErr != nil {
+		m.logger.Error(resultErr)
+
+		return resultErr
+	}
+
+	if rowsAffected == 0 {
+		noRowsErr := errors.New(
+			`database operation failed: no rows match given key and eTag`)
+		m.logger.Error(noRowsErr)
+
+		return noRowsErr
+	}
+
+	if rowsAffected > n {
+		tooManyRowsErr := fmt.Errorf(
+			`database operation failed: more than %d row affected, expected %d, actual %d`,
```

Comment:
nitpick: probably `database operation error:` ? Since the operation did not fail but rather it returned more rows than expected.

Reply:
Ah. I can change.

Reply:
Just a heads up I copied that code from the PostgreSQL state store so the same issue is there as well.